### PR TITLE
Allow admin to add parts with variant selection deferred

### PR DESCRIPTION
**Pull request by dfreerksen**
We have a need to allow users to select a variant for a bundle at order-time rather than having the variant be pre-determined when set up by the admin. A common use-case for this would be choosing a t-shirt size or color.

/cc @huoxito and @JDutil for any concerns with the technical approach.

### What's new

- Introduce PartLineItem model. When a bundle is added to the cart, all pre-determined and user selected variants for a bundle are held here.
- Lots more tests

### What's changed

- Moved JS out of the views and refactored as Coffeescript
- Change in behavior:
  - When a part is already assigned to a bundle, and we use the admin form to add the same part/variant to the bundle with a new quantity the existing `assemblies_part` record was being updated by adding the existing quantity to the newly-supplied one (so if we already had 1, and I enter 1 again, it would end up being 2). This seems confusing from a UX perspective, so now we override the existing quantity

### What's not here

- Pricing change (increase or decrease) based on selected variant.
- Prevent an admin from adding variants to a product if it already has parts.
- Prevent an admin from adding parts to a product if it already has variants.

### CODE REVIEW

File: app/assets/javascripts/spree/backend/spree_product_assembly/index.js.coffee [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    Spree.pathFor("admin/products/" + productSlug + "/parts/available")
+
+  showErrorMessages = (xhr) ->
+    response = JSON.parse(xhr.responseText)
+    show_flash("error", response)
+
+  partsTable = $("#product_parts")
+  searchResultsTable = $("#search_hits")
+
+  search_for_parts = ->
```
Comment (mcmire) [inline app/assets/javascripts/spree/backend/spree_product_assembly/index.js.coffee]: `searchForParts`?

File: app/assets/javascripts/spree/backend/spree_product_assembly/index.js.coffee [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      search_for_parts()
+      false
+    else
+      true
+
+  $("#search_parts_button").click (e) ->
+    e.preventDefault()
+    search_for_parts()
+
+  make_post_request = (link, post_params) ->
```
Comment (mcmire) [inline app/assets/javascripts/spree/backend/spree_product_assembly/index.js.coffee]: `makePostRequest`?

Comment (mcmire) [inline app/assets/javascripts/spree/backend/spree_product_assembly/index.js.coffee]: Perhaps make `post_params` (`postParams`) be optional, with `{}` as default?

File: app/controllers/spree/admin/parts_controller.rb [hunk 3]
```diff
@@ -32,18 +30,34 @@ def available
...
-    render 'spree/admin/parts/update_parts_table'
+    save_part(new_part_params)
   end
 
   private
-    def find_product
-      @product = Spree::Product.find_by(slug: params[:product_id])
+
+    def save_part(part_params)
+      form = Spree::AssignPartToBundleForm.new(product, part_params)
```
Comment (mcmire) [inline app/controllers/spree/admin/parts_controller.rb]: Cool, love the form object.

File: app/models/spree/assign_part_to_bundle_form.rb [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    def variant_selection_deferred?
+      part_options[:variant_selection_deferred]
+    end
+
+    def quantity
+      part_options[:count].to_i
+    end
+
+    def assemblies_part
```
Comment (mcmire) [inline app/models/spree/assign_part_to_bundle_form.rb]: This seems a little misleading -- once it hits one branch it'll stay in that branch. What about:

``` ruby
@assemblies_part ||= begin
  if given_id?
    ...
  else
    ...
  end
end
```

Or, alternatively, don't memoize this here and instead set the instance variable in the constructor.

File: app/views/spree/admin/parts/available.html.erb [hunk 5]
```diff
@@ -44,13 +41,3 @@
     <% end %>
   </tbody>
 </table>
-
-<%= javascript_tag do %>
```
Comment (mcmire) [inline app/views/spree/admin/parts/available.html.erb]: :+1:

File: app/views/spree/admin/parts/index.html.erb [hunk 6]
```diff
@@ -1,7 +1,11 @@
+<%- content_for :javascript do %>
+  <%= javascript_tag "backend/spree_product_assembly" %>
```
Comment (mcmire) [inline app/views/spree/admin/parts/index.html.erb]: Does this not get included automatically for some reason?

File: spec/features/admin/parts_spec.rb [hunk 7]
```diff
@@ -4,25 +4,182 @@
...
+          name: product_name,
+          option_types: [option_type]
+        )
+        create(:variant,
+          product: product,
+          option_values: option_type.option_values
+        )
+      end
+
+     it "when a specific variant is selected" do
```
Comment (mcmire) [inline spec/features/admin/parts_spec.rb]: Did you mean for this to be a `context`, or rename this test?

File: app/views/spree/admin/parts/available.html.erb [hunk 8]
```diff
@@ -34,7 +30,8 @@
 		    <td class="actions">
           <%= image_tag "spinner.gif", :style => "display:none", :class => "spinner" %>
 		      <%= link_to(icon('add') + ' ' + Spree.t(:select), 
-		                  admin_product_parts_path(@product),
+		                  admin_product_parts_path(product),
+                      :data => { 'target' => dom_id(part), 'master-variant-id' => part.master.id},
```
Comment (mcmire) [inline app/views/spree/admin/parts/available.html.erb]: Align this with `:class`?

File: app/assets/javascripts/spree/backend/spree_product_assembly/index.js.coffee [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    else
+      part.variant_id = $('input[name="part[id]"]', row).val()
+
+    make_post_request(link, {assemblies_part: part})
+
+  partsTable.on "click", "a.set_count_admin_product_part_link", ->
+    params = { count :  $("input", $(this).parent().parent()).val() }
+    make_post_request($(this), params)
+
+  partsTable.on "click", "a.remove_admin_product_part_link",  ->
```
Comment (vinniefranco) [inline app/assets/javascripts/spree/backend/spree_product_assembly/index.js.coffee]: Is that two spaces :eyes: spy?

File: app/assets/javascripts/spree/backend/spree_product_assembly/index.js.coffee [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        part.variant_selection_deferred = "t"
+        part.variant_id = link.data("master-variant-id")
+
+    else
+      part.variant_id = $('input[name="part[id]"]', row).val()
+
+    make_post_request(link, {assemblies_part: part})
+
+  partsTable.on "click", "a.set_count_admin_product_part_link", ->
+    params = { count :  $("input", $(this).parent().parent()).val() }
```
Comment (vinniefranco) [inline app/assets/javascripts/spree/backend/spree_product_assembly/index.js.coffee]: Two spaces after ":"?

Comment (mcmire) [inline app/assets/javascripts/spree/backend/spree_product_assembly/index.js.coffee]: Also one space after `count`

File: app/assets/javascripts/spree/backend/spree_product_assembly/index.js.coffee [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+    part = {}
+    link = $(this)
+    row = $("#" + link.data("target"))
+    loadingIndicator = $("img.spinner", link.parent())
+    quantityField = $('input:last', row)
+
+    part.count = quantityField.val()
+
+    if (row.hasClass("with-variants"))
```
Comment (mcmire) [inline app/assets/javascripts/spree/backend/spree_product_assembly/index.js.coffee]: You don't need the parens here.

File: app/assets/javascripts/spree/backend/spree_product_assembly/index.js.coffee [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+      selectedVariantOption = $('select option:selected', row)
+      part.variant_id = selectedVariantOption.val()
+
+      if (selectedVariantOption.text() == Spree.translations.user_selectable)
```
Comment (mcmire) [inline app/assets/javascripts/spree/backend/spree_product_assembly/index.js.coffee]: You don't need the parens here either.

File: app/assets/javascripts/spree/backend/spree_product_assembly/index.js.coffee (hunk 1 above)
Comment (anthonynavarre) [inline app/assets/javascripts/spree/backend/spree_product_assembly/index.js.coffee]: Yeah, underscore-cased variable names are a hold-over from the original JS — good catch

**Comment by yf6** [comment]
Hey guys, any updates on this PR? This is an awesome feature and it would be cool if it can get merged.

**Comment by vinniefranco** [comment]
@yf6 Whoa. Slipped our minds! We'll get it in ASAP

**Comment by yf6** [comment]
:+1:

Yudi

**Comment by yf6** [comment]
bump?

**Comment by jsilland** [comment]
@dfreerksen, @mcmire — short of the CI failing, is there any more issues with this PR that need to be addressed before it can be merged in? We need the functionality brought by this commit and would like to get the ball rolling again.

**Comment by vinniefranco** [comment]
@jsilland We've been waiting for whomever owns this repo to accept the PR for months. Perhaps ground(ctrl) should be the primary for this gem?

**Comment by jsilland** [comment]
@vinniefranco This repo belongs to an organization, I don't think org subrepos allow for designated owners. If there's no progress on this by Monday evening PST, I'll start pinging owners of spree-contrib directly

**Comment by vinniefranco** [comment]
Oh I know that. I was just suggesting that people use our fork, since this
appears to be stagnant.

On Friday, July 17, 2015, Julien Silland <notifications@github.com> wrote:

> @vinniefranco <https://github.com/vinniefranco> This repo belongs to an
> organization, I don't think org subrepos allow for designated owners. If
> there's no progress on this by Monday evening PST, I'll start pinging
> owners of spree-contrib directly
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spree-contrib/spree-product-assembly/pull/109#issuecomment-122459537>
> .
>


-- 
Vincent Franco

"Any intelligent fool can make things bigger and more complex... It takes a
touch of genius --- and a lot of courage to move in the opposite
direction."   -- Einstein

**Comment by futhr** [comment]
Hi guys, I have not been working with this repository for some time. @huoxito used to be the main guy behind it but I assume he has been very busy elsewhere. You need to sort out the failing Travis build before it can be merged.

**Comment by futhr** [comment]
I just noticed that the `master` branch of this repository is still using Spree 2.4.x with 3 major PR's to fix it.

**Comment by huoxito** [comment]
yup I'm aware people have been looking forward to have this for a long while but I haven't really been doing anything product-assembly related so I dont feel safe merging this.

btw I guess this project still needs a maintainer and we're willing to give access to those interested

**Comment by jsilland** [comment]
@huoxito — @vinniefranco and the organization he belongs to (ground(ctrl)) seem to be the best candidates for this. Is there anything standing in the way of granting them ownership?

**Comment by huoxito** [comment]
@vinniefranco would you like to have commit access here so you guys can take over?

I'd email you but I don't see your email on your github profile. The only thing we ask is to not push new feature code directly to master, submit a PR before to get feedback and all. Then you could apply further patches and merge this once the build is green.

**Comment by vinniefranco** [comment]
Sounds good to me. We'd be happy to maintain this for you

On Thursday, July 23, 2015, Washington L Braga Jr <notifications@github.com>
wrote:

> @vinniefranco <https://github.com/vinniefranco> would you like to have
> commit access here so you guys can take over?
>
> I'd email you but I don't see your email on your github profile. The only
> thing we ask is to not push new feature code directly to master, submit a
> PR before to get feedback and all. Then you could apply further patches and
> merge this once the build is green.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spree-contrib/spree-product-assembly/pull/109#issuecomment-124162882>
> .
>


-- 
Vincent Franco

"Any intelligent fool can make things bigger and more complex... It takes a
touch of genius --- and a lot of courage to move in the opposite
direction."   -- Einstein

**Comment by vinniefranco** [comment]
@huoxito just checking in here, let me know if we are good to go with adding us as a maintainer

**Comment by huoxito** [comment]
ops I had added you as a commiter to this repo @vinniefranco but forgot to mention here, currently it says pending invitation, did you check your email?

screen shot 2015-08-03 at 2 43 37 pm

**Comment by vinniefranco** [comment]
@huoxito Weird. I don't have any invites... cancel and resend?

**Comment by huoxito** [comment]
sure @vinniefranco just invited you again

**Comment by vinniefranco** [comment]
:+1:  @jsilland @yf6

**Comment by yf6** [comment]
:confetti_ball:

**Pull request merged**